functions/logStripeEvent.js

```diff
@@ -103,10 +126,10 @@ exports.handlePaymentIntentSucceeded = async (intent, id) => {
         console.warn(`event ${id} is missing param ${param}`);
       }
     }
-    const uid = await helpers.getOrCreateDonor(params);
+    const uid = get(existingDonation, 'sourceDonor', await helpers.getOrCreateDonor(params));
     params['sourceDonor'] = uid;
     console.log(`user is ${uid}`);
-    logDonation.writeDonation(params); // kick off the asynchronous write
+    await logDonation.writeDonation(params, existingDonation); // kick off the asynchronous write
```

Comment:
adding this `await` will significantly increase the response time of `logPaymentIntent`, since `writeDonation` has a lot of async work being done in it. We don't pass any information back upstream from this function, so is there another reason why `await`ing this function is a good idea?

Reply:
If we don't await that method, Stripe can receive a 200 response, believing that the webhook executed successfully, when in reality it could fail after the response is sent to Stripe.  Stripe will retry the webhook on failure (which is what we want, in case there was a timeout with the DB or something outside of the code failed) and ideally save the payment the second call.  

What I would do is pull out the learner assignment and email notification calls from the `writeDonation` function (two reasons - those two actions are outside the goal of writing a donation to the DB & so we can choose to await some/all of them).  We can then await the actual logic of writing a donation to the DB (something that Stripe should wait on since it's related to the webhook information being saved in our system) and then decide if we also want to await the learner assignment and notification email being sent.

We probably don't have to wait for the email to be sent IMO since that's outside the flow for saving a payment but I feel like we should await user assignment.  We also need to determine how GCF's handle a child thread when the parent thread completes (some frameworks will let the child thread silently fail, so we wouldn't have any error logs).  

I'll update the code with those two function calls pulled out.  Do you know how much longer it takes with the await?  I'm thinking that since it's a webhook calling the endpoint, another 1-2 seconds shouldn't affect anything

Reply:
I've seen it take longer than that. More than 10 seconds, sometimes as much as a minute. It's worth looking into whether that's just the latency of the firestore API or if we could optimize our code to minimize "cold starts" of our GCF instances

Reply:
After making updates to the code based on your comments, the code is not awaiting the email notification but it is awaiting the write of the donation to Firebase and the learner assignment.  Creating a new donation and assigning users is taking ~1.5 seconds and a replay event is taking ~500ms.  

The program continues after the HTTP response is sent, to send the email to the user.  There shouldn't be any issues with skipping the await for the email notification (the longest running task).  

Thanks for pointing out that it was taking a long time to run all of the async code!